### PR TITLE
Correct internal gesture ID for Seika Notetaker

### PR DESCRIPTION
**Pull request by school510587**
### Link to issue number:

Fixes issue 3 of #16828 .

### Summary of the issue:

The br(seikantk):XXX IDs of bk:space and bk:space+dots are incorrect. Although it has few influence to UX, it may bring trouble to add-on developers.

### Description of user facing changes

There must be one option displayed when adding a new gesture in Input gestures dialog is about Seika Notetaker. Without this PR, the option is always `backspace+dX`, e.g. `backspace+d1+d3+d4+d5`. Now, `backspace+d1+d3+d4+d5`, `d1+d3+d4+d5+space`, and `backspace+d1+d3+d4+d5+space` are displayed correctly according to the actual space key(s) pressed by the user.

### Description of development approach

Type of `space` parameter of class InputGesture becomes int. It may be 1(backspace), 2(space), or 3(both). The correct key name is obtained from `_getKeyNames`.

### Testing strategy:

1. Open `Input gestures` dialog.
2. Select an item and press Add.
3. Press (the right) space with dots 1, 3, 4, and 5 on Seika Notetaker.
4. `d1+d3+d4+d5+space (Seika Notetaker)` should be present.

If both spaces are pressed, the result should be `backspace+d1+d3+d4+d5+space`.

### Known issues with pull request:

3 entries will be required to build a new item in gesture map. For example, `br(seikantk):BACKSPACE+d1+d3+d4+d5`, `br(seikantk):d1+d3+d4+d5+SPACE`, and `br(seikantk):BACKSPACE+d1+d3+d4+d5+SPACE` must be present to completely define the action for space with dots 1, 3, 4, and 5.

### Code Review Checklist:

- [ ] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [ ] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] API is compatible with existing add-ons.
- [ ] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
	- Enhanced gesture handling logic to improve user interaction with the input system.
	- Updated space parameter representation to reflect key values, improving clarity and functionality.
- **Bug Fixes**
	- Ensured consistency in the interpretation of the space parameter across different methods.

### CODE REVIEW

**Review by SaschaCowley — CHANGES_REQUESTED** [review]
This seems to be fine to me, though I don't have a Seika Notetaker to test with. I just have a question about the internal determination of the space key which I would like clarification on (I understand you haven't changed the lines of code I'm asking about, but I reckon we can clarify this small matter while we're here). Also, could you please write a changelog entry for this fix?

**Comment by SaschaCowley** [comment]
@school510587 have you thoroughly tested this yourself?

**Comment by school510587** [comment]
Hi @SaschaCowley,

> Why are we using (1, 2, 3) for space at line 310, but (1, 2) for space at line 316?

`(1, 2, 3)` in line 310 is quite easy to understand. It represents that at least one of BACKSPACE and SPACE are pressed.

`BACKSPACE + SPACE`, i.e. value 3 of line 316, is defined as `sayAll`, rather than ordinary space. So, `brailleDots = 0 and key = 3` fails both tests in line 309 and 316, and finally forms `br(seikantk):BACKSPACE+SPACE` at line 319. It is not a braille input, because both `dots` and `space` properties are not set.

**Comment by school510587** [comment]
Hi @SaschaCowley,

> @school510587 have you thoroughly tested this yourself?

Yes, here is a detailed result.

<p>
<table border="1">
<tr><th>Key</th><th>Input Help</th><th>Input Gesture options</th></tr>
<tr>
<td>Both BACKSPACE and SPACE</td>
<td>SPACE + BACKSPACE</td>
<td>backspace + space (Seika Notetaker)</td>
</tr>
<tr>
<td>Single BACKSPACE</td>
<td>space</td>
<td>backspace (Seika Notetaker), any dots</td>
</tr>
<tr>
<td>Single SPACE</td>
<td>space</td>
<td>space (Seika Notetaker), any dots</td>
</tr>
<tr>
<td>BACKSPACE, SPACE and dot 1</td>
<td>space with dot 1</td>
<td>backspace+d1+space (Seika Notetaker), space with dot 1, space with any dots</td>
</tr>
<tr>
<td>Both BACKSPACE and dot 1</td>
<td>space with dot 1</td>
<td>backspace+d1 (Seika Notetaker), space with dot 1, space with any dots</td>
</tr>
<tr>
<td>Both SPACE and dot 1</td>
<td>space with dot 1</td>
<td>d1+space (Seika Notetaker), space with dot 1, space with any dots</td>
</tr>
<tr>
<td>Single dot 1</td>
<td>dot 1</td>
<td>d1 (Seika Notetaker), dot 1, any dots</td>
</tr>
</table>
</p>

A problem appears above. Key name in the input help is not precise enough, but the following description of its function is correct. That is, if different functions are bound to BACKSPACE and SPACE respectively, they share the same key name `space` with different descriptions in input help.

Another notable phenomenon is that the key name of `BACKSPACE + SPACE` is precise. The difference is that it is not a braille input. So far, I don't know why the key name of braille input gesture is not precise enough.

**Comment by seanbudd** [comment]
cc @moyanming

**Comment by seanbudd** [comment]
cc @moyanming - any thoughts on this PR and the issues shown in input help?

**Review by seanbudd — APPROVED** [review]
Approved. If we get further feedback from @moyanming or other users/testers/devs we can address issues.
Please make sure this gets entries in the change log under bug fixes and API breaking changes

**Review by SaschaCowley — APPROVED** [review]

**Review by SaschaCowley — COMMENTED** [review]
@school510587, please add changes entries.

**Comment by school510587** [comment]
Hi,

Sorry, which is correct?

`* Correct internal gesture ID for Seika Notetaker. (#16828, @school510587)`

or

`* Correct internal gesture ID for Seika Notetaker. (#17047, @school510587)`

The PR 17047 attempts to solve the known issue 3 of 16828. Both are not issue number, and which should I use? Thanks.

**Comment by seanbudd** [comment]
Either the issue number or this PR number is fine

File: user_docs/en/changes.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ The available options are:
   * It is now possible to use braille display routing keys to move the text cursor. (#9101)
   * It is now possible to use the review cursor selection commands to select text.
 * Updating NVDA while add-on updates are pending no longer results in the add-on being removed. (#16837)
+* Correct internal gesture ID for Seika Notetaker. (#17047, @school510587)
```
Comment (SaschaCowley) [inline user_docs/en/changes.md:46]: This needs to be moved to 2025.1. It also needs to explain the user-facing change.

File: user_docs/en/changes.md
Comment (SaschaCowley) [inline user_docs/en/changes.md:1]: Please add an entry to Changes for developers

**Review by SaschaCowley — CHANGES_REQUESTED** [review]

**Comment by school510587** [comment]
Hi,

What's up? The merge seems to be failed, but all tests passed. What can I do to help this?

File: user_docs/en/changes.md
Comment (SaschaCowley) [reply user_docs/en/changes.md:1]: @school510587 you need to merge master into your branch, then add an entry to the "Bug Fixes" section describing the user-facing changes, and an entry to the "Changes for Developers" section describing what you have changed internally.

**Review by SaschaCowley — CHANGES_REQUESTED** [review]

File: user_docs/en/changes.md [hunk 2]
```diff
@@ -51,6 +52,9 @@ Add-ons will need to be re-tested and have their manifest updated.
   * A `toggleBooleanValue` helper function has been added to `globalCommands`.
   It can be used in scripts to report the result when a boolean is toggled in `config.conf`
 * Removed the requirement to indent function parameter lists by two tabs from NVDA's Coding Standards, to be compatible with modern automatic linting. (#17126, XLTechie)
+* Correct the internal IDs of the braille input gesture objects for Seika Notetaker. (#17047, @school510587)
+  * Parameter space of class InputGesture constructor becomes an integer between 0 and 3.
+  * Gestures involving br(seikantk):BACKSPACE and/or br(seikantk):SPACE can be correctly linked to command scripts.
```
Comment (seanbudd) [inline user_docs/en/changes.md:57]: Sorry, this should be under API breaking changes.
The issue is that you are breaking the API, we don't need a general description of the code changes here.
Please read [this doc](https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/deprecations.md) for more information.

The message should be something like


```py
* The `space` keyword argument for `brailleDisplayDrivers.seikantk.BrailleDisplayDriver` now expects an `int` not a `bool`. (#17047)
```

**Comment by school510587** [comment]
Hi,

However, I have another problem. My previous PR 16828 was accepted in NVDA 2024.4, but the IDs of single space and space with dots are not correct. They are corrected by this PR now, and where should I denote this change? Whether should I need to write this into changes.md?

**Comment by SaschaCowley** [comment]
@school510587 [said](https://github.com/nvaccess/nvda/pull/17047#issuecomment-2352638419):

> However, I have another problem. My previous PR 16828 was accepted in NVDA 2024.4, but the IDs of single space and space with dots are not correct. They are corrected by this PR now, and where should I denote this change? Whether should I need to write this into changes.md?

I am not sure I understand what you are asking. PR #16828 works as expected, doesn't it? The only user facing change here is that the space key in use is displayed correctly, but NVDA's actual behaviour is unchanged.

**Comment by school510587** [comment]
There are several IDs for a braille display gesture. For example, `bk:space+dot1`, `bk:space+dots`, and `br(seikantk):SPACE+d1` refer to the same gesture.
#16828 fixes the behavior in `bk` level, i.e. `bk:space` or `bk:space+dots`. However, the behavior in `br(seikantk)` level is incorrect.

> PR #16828 works as expected, doesn't it?

It does achieve my first goal - to generate the corresponding `bk` IDs, and the default braille input behaviors are also bound to them. The displayed name in input help changes from `SPACE+d1` to `space with dot 1`. However, a user will find that `br(seikantk)` are wrong, i.e. `br(seikantk):BACKSPACE+d1` on input gestures dialog.

> The only user facing change here is that the space key in use is displayed correctly, but NVDA's actual behaviour is unchanged.

In this PR, information of the pressed space key(s) is passed into InputGesture class. Therefore, `br(seikantk)` IDs involving `BACKSPACE` and `SPACE` are fixed, and actions bound into `br(seikantk)` IDs are also mapped correctly. However, I surprisingly find no change in the input help hints of gestures that map to the actions by their `br(seikantk)` IDs.

With the initial NVDA settings, the behavior is unchanged from #16828 to this PR. A user must try to alter some setting through input gestures dialog to perceive what described above

**Comment by SaschaCowley** [comment]
Nevertheless, this is an API breaking change, so has to be in a 20xy.1 release. And if I am understanding correctly, your previous PR works as expected and does not introduce any regressions, so there is no need to change anything there.

In future, if you want changes to be introduced into the same NVDA version, please either note this explicitly in all PRs, or open them as a single PR.

**Comment by school510587** [comment]
> Nevertheless, this is an API breaking change, so has to be in a 20xy.1 release. And if I am understanding correctly, your previous PR works as expected and does not introduce any regressions, so there is no need to change anything there.

Ok, I think this is correct.

> In future, if you want changes to be introduced into the same NVDA version, please either note this explicitly in all PRs, or open them as a single PR.

Got it, thanks.

**Review by seanbudd — APPROVED** [review]

File: user_docs/en/changes.md [hunk 3]
```diff
@@ -30,6 +30,7 @@ In order to use this feature, the application volume adjuster needs to be enable
 * Improvements when editing in Microsoft PowerPoint:
   * Caret reporting no longer breaks when text contains wide characters, such as emoji. (#17006 , @LeonarddeR)
   * Character location reporting is now accurate (e.g. when pressing `NVDA+Delete`. (#9941, @LeonarddeR)
+* Correct IDs of single space and space with dots gestures for Seika Notetaker on the input gestures dialog. (#17047, @school510587)
```
Comment (SaschaCowley) [inline user_docs/en/changes.md:33]:
```suggestion
* When using the Seika Notetaker, space and space with dots gestures are now displayed correctly in the Input Gestures dialog. (#17047, @school510587)
```

**Review by SaschaCowley — APPROVED** [review]

**Pull request merged**